sdks/python/apache_beam/runners/interactive/caching/pcollection_cache.py

```diff
@@ -0,0 +1,105 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import abc
+
+from future.utils import with_metaclass
+
+
+class PCollectionCache(with_metaclass(abc.ABCMeta)):
+
+  @abc.abstractmethod
+  def __init__(self, location, **writer_kwargs):
+    """Initialize PCollectionCache.
+
+    Args:
+      location (str): Location where the cache data should be stored.
+      **writer_kwargs: Arguments to pass to the underlying writer class.
+    """
+    raise NotImplementedError
+
+  @property
+  @abc.abstractmethod
+  def persist(self):
+    """Whether to persist the underlying data when the object goes out of scope.
+
+    Returns:
+      bool: ``True`` if the unerlying data will be persisted once the object
+          goes out of scope, ``False`` otherwise.
+    """
+    raise NotImplementedError
+
+  @property
+  @abc.abstractmethod
+  def timestamp(self):
+    """Return a timestamp indicating the last time this instance was modified.
+
+    Returns:
+      float: A non-negative number indicating the last time the cache was
+          modified. Caches that were modified more recently should return
+          a larger number.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def reader(self, **reader_kwargs):
+    """Return a reader PTransform which can read a PCollection from cache.
+
+    Args:
+      **reader_kwargs: Arguments to pass to the underlying reader class.
+
+    Returns:
+      A source from which we can read a PCollection.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def writer(self):
```

Comment:
My rationale for passing `writer_kwargs` to `__init__` while passing `reader_kwargs` to `read` and `reader` was as follows:

1. We should be able to write to the same cache object repeatedly. One of the main reasons for this is to make the concept of a "cache" transferable for bounded and unbounded PCollections.
2. It does not make sense to write to the same cache object using a different set of arguments. In fact, it is very likely that this will lead to cache corruption (for example, using a different coder or schema). This is why I opted to pass `writer_kwargs` to the constructor rather than to the `writer` method.
3. In (I expect rather rare) cases, the user may want to pass additional arguments to the `read` and `reader` methods. The example I had in mind is someone wishing to read only a subset of the columns from a Parquet file while debugging, but then pass all the columns into the subsequent PTransform. We made this possible by allowing `read` and `reader` to take additional keyword arguments, although I am not 100% happy with the aesthetics of this.
4. In many cases, the arguments passed to the underlying `writer_class` must also be passed to the underlying `reader_class`. The most common examples may be `coder`, `compression_type`, `schema`, etc. In the interim, I decided to control this using the `_reader_passthrough_arguments` attribute, while ultimately it would be nicer if we could infer these from the underlying `reader_class` and `writer_class` objects themselves?

Reply:
Since the PCollectionCache interface is still experimental, I decided to change `write` and `writer` methods to take on an additional `writer_kwargs` argument. While I do not think that it should be used in most (or any) cases, I understand the confusion that is caused by asymmetry between the `read` and `write` methods (and between the `reader` and `writer` methods).

An alternative would be to remove the `reader_kwargs` argument from the `read` and `reader` methods, and that is also probably something that should be considered?